pydantic/class_validators.py

```diff
@@ -244,86 +244,50 @@ def make_generic_validator(validator: AnyCallable) -> 'ValidatorCallable':
-        # assume the first argument was value which has already been removed
-        return wraps(validator)(_generic_validator_basic(validator, sig, set(args)))
+        if first_arg == 'cls':
+            # assume the second argument is value
+            args = args[1:]
+            skip_first = False
+        else:
+            # assume the first argument was value which has already been removed
+            skip_first = True
+
+        return wraps(validator)(_generic_validator(validator, sig, set(args), skip_first))
 
 
 def prep_validators(v_funcs: Iterable[AnyCallable]) -> 'ValidatorsList':
     return [make_generic_validator(f) for f in v_funcs if f]
 
 
-all_kwargs = {'values', 'field', 'config'}
+args_order = ('values', 'field', 'config')
+all_kwargs = {*args_order}
+required_args = 2
 
 
-def _generic_validator_cls(validator: AnyCallable, sig: 'Signature', args: Set[str]) -> 'ValidatorCallable':
-    # assume the first argument is value
-    has_kwargs = False
+def _generic_validator(
+    validator: AnyCallable, sig: 'Signature', args: Set[str], skip_first: bool = False
+) -> 'ValidatorCallable':
     if 'kwargs' in args:
-        has_kwargs = True
-        args -= {'kwargs'}
+        args = all_kwargs | args - {'kwargs'}
 
     if not args.issubset(all_kwargs):
+        sig_args = ('cls', 'value', *args_order)[skip_first:]
         raise ConfigError(
             f'Invalid signature for validator {validator}: {sig}, should be: '
-            f'(cls, value, values, config, field), "values", "config" and "field" are all optional.'
+            f'({", ".join(sig_args)}), "values", "config" and "field" are all optional.'
         )
 
-    if has_kwargs:
-        return lambda cls, v, values, field, config: validator(cls, v, values=values, field=field, config=config)
-    elif args == set():
-        return lambda cls, v, values, field, config: validator(cls, v)
-    elif args == {'values'}:
-        return lambda cls, v, values, field, config: validator(cls, v, values=values)
-    elif args == {'field'}:
-        return lambda cls, v, values, field, config: validator(cls, v, field=field)
-    elif args == {'config'}:
-        return lambda cls, v, values, field, config: validator(cls, v, config=config)
-    elif args == {'values', 'field'}:
-        return lambda cls, v, values, field, config: validator(cls, v, values=values, field=field)
-    elif args == {'values', 'config'}:
-        return lambda cls, v, values, field, config: validator(cls, v, values=values, config=config)
-    elif args == {'field', 'config'}:
-        return lambda cls, v, values, field, config: validator(cls, v, field=field, config=config)
-    else:
-        # args == {'values', 'field', 'config'}
-        return lambda cls, v, values, field, config: validator(cls, v, values=values, field=field, config=config)
-
+    def _fetch_args(args_: Tuple[Any, ...]) -> Dict[str, Any]:
+        return {k: args_[args_order.index(k)] for k in args_order if k in args}
```

Comment:
This is just a guess but could it be that @samuelcolvin used the big lambda list for performance reasons since fetch args is now run on every call of the validator?  

A side anecdote, while working on another pull request I have to say that this does make development MUCH easier... I was taking for ever to make all the changes in the correct places and since the validator code swallows the errors it's hard to track down the exact place where stuff fails, so if there aren't performance or editor/ide support reasons keeping the old versions I'm all for using your version here.

Reply:
@daviskirk I faced the same problem when was debugging an issue with `pydantic`, it's nice to have one entry point to call validator functions.

But I should say that code with lambdas was more explicit rather than my version.

It will be great to hear @samuelcolvin opinion about this PR🙂

Reply:
Fyi @uriyyo I was using your code in https://github.com/samuelcolvin/pydantic/pull/2018 and went ahead and checked the benchmarks `python benchmarks/run.py` or equivalent make command and they were indeed worse than the baseline using master.
That said I might have slowed things down with the other changes I made, just wanted to say that the benchmarks are there and run if you want to do a comparison on this code as well.

Reply:
@daviskirk Thanks for such important information.
Looks like I found places which lent for performance issues.

After update, if compare my branch and current master branch, benchmark results are:
my branch - pydantic best=50.963μs/iter avg=52.860μs/iter stdev=3.121μs/iter
master - pydantic best=49.436μs/iter avg=50.348μs/iter stdev=0.712μs/iter